hal/src/photon/concurrent_hal.cpp

```diff
@@ -114,11 +103,239 @@ os_result_t os_thread_join(os_thread_t* thread)
  * @param thread    The thread to clean up.
  * @return 0 on success.
  */
-os_result_t os_thread_cleanup(os_thread_t* thread)
+os_result_t os_thread_cleanup(os_thread_t thread)
 {
-    return wiced_rtos_delete_thread((wiced_thread_t*)thread);
+    vTaskDelete(thread);
+    return 0;
 }
 
+/**
+ * Map gthread handles to FreeRTOS task handles.
+ */
+typedef TaskHandle_t __gthread_t;
+
+
+class ThreadQueue
+{
+    QueueHandle_t queue;
+
+public:
+
+    ThreadQueue(UBaseType_t max_size) {
+        queue = xQueueCreate(max_size, sizeof(TaskHandle_t));
+    }
+
+    ~ThreadQueue() {
+        vQueueDelete(queue);
+    }
+
+
+    void enqueue(TicksType_t ticks=portMAX_DELAY) {
+        TaskHandle_t current = xTaskGetCurrentTaskHandle();
+        xQueueSend(queue, &current, ticks);
+    }
+
+    void sleep() {
+        enqueue();
+        vTaskSuspend(NULL);
+    }
+
+    bool wake() {
+        TaskHandle_t next;
+        if (xQueueReceive(queue, &next, 0)==pdTRUE) {
+            vTaskResume(next);
+            return true;
+        }
+        return false;
+    }
+
+    void wakeAll()
+    {
+        while (wake()) {}
+    }
+};
+
+
+class Lock
+{
+    SemaphoreHandle_t semaphore;
+
+public:
+
+    Lock() {
+        semaphore = xSemaphoreCreateMutex();
+    }
+
+    void acquire() {
+        while (xSemaphoreTake(semaphore, portMAX_DELAY)==pdFALSE);
+    }
+
+    void release() {
+        xSemaphoreGive(semaphore);
+    }
+
+    ~Lock() {
+        vSemaphoreDelete(semaphore);
+    }
+};
+
+
+class ConditionVariable
+{
+    ThreadQueue queue;
+
+    typedef std::unique_lock<std::mutex> lock_t;
+public:
+    ConditionVariable(UBaseType_t max_size) : queue(max_size) {}
+
+    void wait(lock_t* lock)
+    {
+        taskENTER_CRITICAL();
+        lock->unlock();
+        queue.enqueue();
+        taskEXIT_CRITICAL();
+        vTaskSuspend(NULL);
+        lock->lock();
+    }
```

Comment:
Is it possible/realistic/straightforward to restructure the code so there
is not even the whiff of a deadlock, rather than debating the
presence/absence of one ?

On Fri, Sep 18, 2015 at 1:53 PM, Matthew McGowan <notifications@github.com>
wrote:

> In hal/src/photon/concurrent_hal.cpp
> <https://github.com/spark/firmware/pull/614#discussion_r39888667>:
>
> > +{
> > +    ThreadQueue queue;
> > +
> > +    typedef std::unique_lock<std::mutex> lock_t;
> > +public:
> > +    ConditionVariable(UBaseType_t max_size) : queue(max_size) {}
> > +
> > +    void wait(lock_t* lock)
> > +    {
> > +        taskENTER_CRITICAL();
> > +        lock->unlock();
> > +        queue.enqueue();
> > +        taskEXIT_CRITICAL();
> > +        vTaskSuspend(NULL);
> > +        lock->lock();
> > +    }
>
> The mutex is provided by the lock used with std::condition_variable. Even
> so, without the lock it's a typical race condition between two threads - if
> vTaskResume is called before vTaskSuspend then that task simply misses the
> notification - just as if it called wait a few microseconds later. I'm not
> seeing a deadlock here, just a missed notification, and the guard against
> that is the lock owned by threads using wait and notify on the condition
> variable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark/firmware/pull/614/files#r39888667>.
>



-- 
Andy